apps/omg_watcher/lib/exit_processor/core.ex

```diff
@@ -507,6 +587,190 @@ defmodule OMG.Watcher.ExitProcessor.Core do
     {chain_validity, events}
   end
 
+  @spec get_output_challenge_data(ExitProcessor.Request.t(), t(), Signed.tx_bytes(), 4..7) :: %{
+          in_flight_txbytes: Transaction.tx_bytes(),
+          in_flight_output_pos: Utxo.Position.t(),
+          in_flight_tx_inclusion_proof: <<_::_*256>>,
+          spending_tx_bytes: Transaction.tx_bytes(),
+          spending_input_index: 0..3,
+          spending_sig: <<_::520>>
+        }
+  def get_output_challenge_data(
+        %ExitProcessor.Request{blocks_result: _blocks, piggybacked_blocks_result: _piggybacked_blocks} = _request,
+        %__MODULE__{in_flight_exits: _ifes} = _state,
+        _txbytes,
+        _output_index
+      ) do
+    #   "in_flight_txbytes": "F3170101C0940000...",
+    #   "in_flight_output_pos": 21000634002,
+    #   "in_flight_proof": "03F451067A805540000...",
+    #   "spending_txbytes": "F847010180808080940000...",
+    #   "spending_input_index": 1,
+    #   "spending_sig": "9A23010180808080940000..."
+  end
+
+  @spec get_input_challenge_data(ExitProcessor.Request.t(), t(), Signed.tx_bytes(), 0..7) ::
+          {:ok, input_challenge_data() | output_challenge_data()} | {:error, atom()}
+  def get_input_challenge_data(
+        request,
+        %__MODULE__{in_flight_exits: ifes} = state,
+        txbytes,
+        input_index
+      ) do
+    with {:ok, tx} <- Transaction.decode(txbytes),
+         {:known_ife, true} <- {:known_ife, Transaction.hash(tx) in Map.keys(ifes)},
+         {:ok, proofs} <- produce_proofs_and_find_the_requested_one(request, state, tx, input_index) do
+      {:ok, proofs}
+    else
+      {:known_ife, false} -> {:error, :unknown_ife}
+      error -> error
+    end
+  end
+
+  defp produce_proofs_and_find_the_requested_one(
+         %ExitProcessor.Request{blocks_result: blocks, piggybacked_blocks_result: piggybacked_blocks},
+         state,
+         tx,
+         input_index
+       ) do
+    encoded_tx = Transaction.encode(tx)
+    known_txs = get_known_txs(Enum.uniq(piggybacked_blocks ++ blocks)) ++ get_known_txs(state)
+
+    compute_proofs = fn ->
+      get_invalid_piggybacks_on_inputs(known_txs, state)
+      |> Enum.filter(fn {ife_encoded_tx, _, _, _} ->
+        ife_encoded_tx == encoded_tx
+      end)
+    end
+
+    with {:found_conflicts, [{_encoded_tx, bad_inputs, bad_outputs, proofs}]} <- {:found_conflicts, compute_proofs.()},
+         {:found_pb, true} <- {:found_pb, input_index in (bad_inputs ++ bad_outputs)} do
+      challenge_data = encode_proofs(tx, input_index, proofs)
+      {:ok, challenge_data}
+    else
+      {:found_conflicts, []} -> {:error, :no_double_spent_inputs_in_requested_ife}
+      {:found_pb, false} -> {:error, :no_double_spent_on_particular_input}
+    end
+  end
+
+  defp encode_proofs(tx, input_index, proofs) when input_index in 0..3 do
```

Comment:
name: be specific it's about piggybacks.

Same comment for `produce_proofs_and_find_the_requested_one`

Reply:
done